File: Java/_1334.java
class Solution {
    int n;
    int[][] weight;
    int threshold;
    List<Integer>[] adjList;
    public int findTheCity(int n, int[][] edges, int threshold) {
        this.n = n;
        this.threshold = threshold;
        weight = new int[n][n];
        for (int[] row: weight) Arrays.fill(row, Integer.MAX_VALUE);
        adjList = new ArrayList[n];
        
        for (int i = 0; i < n; i++) {
            adjList[i] = new ArrayList<>();
        }
        
        for (int[] edge : edges) {
            weight[edge[0]][edge[1]] = weight[edge[1]][edge[0]] = edge[2];
            adjList[edge[0]].add(edge[1]);
            adjList[edge[1]].add(edge[0]);
        }

        int[] nInThreshold = new int[n];
        for (int i = 0; i < n; i++) {
            nInThreshold[i] = dijkstra(i);
        }

        return IntStream.range(0, nInThreshold.length)
            .reduce((a, b) -> nInThreshold[a] < nInThreshold[b] ? a : b)
            .getAsInt();
    }
    
    public int dijkstra(int root) {
        int[] minCost = new int[n];
        Arrays.fill(minCost, Integer.MAX_VALUE);
        minCost[root] = 0;
        
        PriorityQueue<int[]> heap = new PriorityQueue<>((a,b) -> Integer.compare(a[0],b[0]));
        heap.add(new int[] {0, root});
        
        while (!heap.isEmpty()) {
            int[] costNNode = heap.poll();
            int cost = costNNode[0], node = costNNode[1];
            if (minCost[node] < cost) continue;
            
            for (int nbor : adjList[node]) {
                int pathCost = costNNode[0] + weight[node][nbor];
                
                if (pathCost < minCost[nbor]) {
                    minCost[nbor] = pathCost;
                    heap.add(new int[] {pathCost, nbor});
                }
            }
        }
        return (int) Arrays.stream(minCost).filter(i-> i <= threshold).count();                    
    }
}

File: Python3/_1334.py
# Dijkstra's uses same procedure as Prim's except while Prim's calculates
# cost of new node as marginal/additional cost from intermediate already 
# calculated, Dijkstra calculates cost as total cost from root itself. 
# Prim's calculates combined cost of nodes as tree, while Dijkstra's 
# calculates cost individually as own path

class Solution:
    def findTheCity(self, n: int, edges: List[List[int]], threshold: int) -> int:

        # inf entry represents no direct edge between two nodes
        weight = [[float('inf')] * n for _ in range(n)]
        
        # Stores list of nbors as val for each node
        adjList = collections.defaultdict(list)
        
        # Stores weight of direct edge between two nodes
        for (a, b, c) in edges:
            weight[a][b] = weight[b][a] = c
            adjList[a].append(b)
            adjList[b].append(a)
            
        def dijkstra(root):
            minCost = [float('inf')] * n
            minCost[root] = 0

            # heap stores cost to node through intermediate to which
            # root is connected (path is not necessarily shortest)

            # cost must be 0th index for heap to sort along cost
            heap = [(0, root)]

            while heap:
                cost, node = heapq.heappop(heap)

                # Dikstra gets greedy. Wants cost of this path to be
                # lower than any other path to that same node
                if minCost[node] < cost: continue

                for nbor in adjList[node]:
                    # cost(root to nbor through node) = 
                    # minCost(root to node, current) + weight(node to nbor)
                    pathCost = cost + weight[node][nbor]

                    # Lower cost along this path than any prev calculated
                    if pathCost < minCost[nbor]:
                        minCost[nbor] = pathCost
                        heapq.heappush(heap, (pathCost, nbor))

            # -1 because each node should exclude itself as nbor
            return sum(cost <= threshold for cost in minCost) - 1
        
        # Iterate backward because in case of tie, want higher index
        nInThreshold = {i : dijkstra(i) for i in range(n - 1, -1, -1)}
        return min(nInThreshold, key = nInThreshold.get)